src/client/shadows/dynamicshadowsrender.cpp

```diff
@@ -177,21 +177,24 @@ void ShadowRenderer::setShadowIntensity(float shadow_intensity)
 void ShadowRenderer::addNodeToShadowList(
 		scene::ISceneNode *node, E_SHADOW_MODE shadowMode)
 {
-	if (!node)
-		return;
 	m_shadow_node_array.emplace_back(node, shadowMode);
+	// If node is clientMap, cache its ID
+	if (strcmp(node->getName(), "ClientMap") == 0)
+		m_shadow_node_clientmap_cache_id = m_shadow_node_array.size() - 1;
+
 	if (shadowMode == ESM_RECEIVE || shadowMode == ESM_BOTH)
 		node->setMaterialTexture(TEXTURE_LAYER_SHADOW, shadowMapTextureFinal);
 }
 
 void ShadowRenderer::removeNodeFromShadowList(scene::ISceneNode *node)
 {
-	if (!node)
-		return;
 	node->setMaterialTexture(TEXTURE_LAYER_SHADOW, nullptr);
 	for (auto it = m_shadow_node_array.begin(); it != m_shadow_node_array.end();) {
 		if (it->node == node) {
 			it = m_shadow_node_array.erase(it);
+			// If node is clientMap, reset our cached ID
+			if (strcmp(node->getName(), "ClientMap") == 0)
+				m_shadow_node_clientmap_cache_id = -1;
```

Comment:
if you remove something from the array the clientmap might change position, this is not taken into account

Reply:
exact

Reply:
fixed by removing this caching and changing design